common/src/borrow.rs

```diff
@@ -18,6 +18,31 @@ pub enum BorrowedValue<'a, T: ?Sized> {
     MappedReadLock(PyMappedRwLockReadGuard<'a, T>),
 }
 
+impl<'a, T: ?Sized> BorrowedValue<'a, T> {
+    pub fn map<U: ?Sized, F>(s: Self, f: F) -> BorrowedValue<'a, U>
+    where
+        F: FnOnce(&T) -> &U,
+    {
+        match s {
+            Self::Ref(r) => BorrowedValue::Ref(f(r)),
+            Self::MuLock(m) => BorrowedValue::MappedMuLock(PyMutexGuard::map(m, |x| unsafe {
+                #[allow(mutable_transmutes, clippy::transmute_ptr_to_ptr)]
+                std::mem::transmute(f(x))
```

Comment:
This isn't sound -- somebody could match to get the mutex guard out of the enum, and then treat it as mutable. I ran into the same problem, and I'm not sure exactly what the solution is. Maybe try something in cell? ImmutableMappedMutexGuard or something?

Reply:
I could try making that later today if you don't feel confident about it; I think it would be like `{ data: *const T, raw: &'a parking_lot::RawMutex, _marker: PhantomData<(&'a T, RawMutex::Guard market)> }`

Reply:
If you can fix that will be perfect, I have no clue to get it soundless..

Reply:
#2239

Reply:
@coolreader18 will you merge #2239 before or after the pr?